savant/api/parser.py

```diff
@@ -0,0 +1,139 @@
+from typing import Union
+
+from savant_rs.primitives import (
+    Attribute,
+    AttributeValue,
+    AttributeValueType,
+    VideoFrame,
+    VideoObject,
+)
+from savant_rs.primitives.geometry import BBox, RBBox
+from savant_rs.video_object_query import MatchQuery
+
+from savant.meta.constants import UNTRACKED_OBJECT_ID
+
+
+def parse_video_frame(frame: VideoFrame):
+    # TODO: add content to metadata if its not embedded
+    return {
+        'source_id': frame.source_id,
+        'framerate': frame.framerate,
+        'width': frame.width,
+        'height': frame.height,
+        'pts': frame.pts,
+        'keyframe': frame.keyframe,
+        'codec': frame.codec,
+        'dts': frame.dts,
+        'duration': frame.duration,
+        'metadata': {'objects': parse_video_objects(frame)},
+        'tags': parse_tags(frame),
+    }
+
+
+def parse_tags(frame: VideoFrame):
+    return {
+        name: parse_attribute_value(frame.get_attribute(namespace, name).values[0])
+        for namespace, name in frame.attributes
+    }
+
+
+def parse_video_objects(frame: VideoFrame):
+    parents = {}
+    objects = {}
+    for obj in frame.access_objects(MatchQuery.idle()):
+        for child in frame.get_children(obj.id):
+            parents[child.id] = obj
+        objects[obj.id] = parse_video_object(obj)
+
+    for obj_id, parent in parents.items():
+        child = objects[obj_id]
+        child['parent_model_name'] = parent.namespace
+        child['parent_label'] = parent.label
+        child['parent_object_id'] = parent.get_track_id()
+
+    return list(objects.values())
+
+
+def parse_video_object(obj: VideoObject):
+    track_id = obj.get_track_id()
+    if track_id is None:
+        track_id = UNTRACKED_OBJECT_ID
+
+    return {
+        'model_name': obj.namespace,
+        'label': obj.label,
+        'object_id': track_id,
+        'bbox': parse_bbox(obj.detection_box),
+        'confidence': obj.confidence,
+        'attributes': [
+            parse_attribute(obj.get_attribute(namespace, name))
+            for namespace, name in obj.attributes
+        ],
+        'parent_model_name': None,
+        'parent_label': None,
+        'parent_object_id': None,
+    }
+
+
+def parse_attribute(attribute: Attribute):
+    value = attribute.values[0]
+    return {
+        'element_name': attribute.namespace,
+        'name': attribute.name,
+        'value': parse_attribute_value(value),
+        'confidence': value.confidence,
+    }
+
+
+def parse_attribute_value(value: AttributeValue):
```

Comment:
@tomskikh I made `AttributeValueType` hashable in `0.1.38`.

Reply:
Updated savant-rs to 0.1.38